### ISSUE TITLE
Remove shadow properties

### ISSUE THREAD

**Issue by loic-sharma** [loic-sharma/BaGet#445]
## Bug

If BaGet is started in debug mode it will log:

```
dbug: Microsoft.EntityFrameworkCore.Model.Validation[10600]
      The property 'PackageKey' on entity type 'PackageDependency' was created in shadow state because there are no eligible CLR members with a matching name.
dbug: Microsoft.EntityFrameworkCore.Model.Validation[10600]
      The property 'PackageKey' on entity type 'PackageType' was created in shadow state because there are no eligible CLR members with a matching name.
dbug: Microsoft.EntityFrameworkCore.Model.Validation[10600]
      The property 'PackageKey' on entity type 'TargetFramework' was created in shadow state because there are no eligible CLR members with a matching name.
```

These logs were added in EF Core 3.

## Fix

1. Add a `public int PackageKey { get; set; }` property to the `PackageType`, `PackageDependency`, and `TargetFramework` entities.
1. Generate new migrations, if any

## Additional Context

See the documentation on shadow properties: https://docs.microsoft.com/en-us/ef/core/modeling/shadow-properties#foreign-key-shadow-properties

**Comment by MarkZither** [comment]
I am working on this, having read the link regarding shadow properties and then [relationships](https://docs.microsoft.com/en-us/ef/core/modeling/relationships?tabs=fluent-api%2Cfluent-api-simple-key%2Csimple-key) i added the properties `public int PackageKey { get; set; }` to `PackageDependency`, `PackageType` and `TargetFramework`.

This works and stops the messages being logged, however it has made the FK relationship non-nullable therefore producing bad migrations. Which I see in the comment in `AbstractContext` is desired but isn't possible with sqlite migrations, so for now  I have changed it to `public int? PackageKey { get; set; }` which then does not produce any changes in the migration.

Based on this [discussion in the ef repo](https://github.com/dotnet/efcore/issues/329#issuecomment-276138743) I am trying to produce a migration to copy the PackageDependencies data to a new table with the new non-nullable FK, but that might be better as a separate issue?

**Comment by loic-sharma** [comment]
Thanks for the help once again, I really appreciate it :)

> This works and stops the messages being logged, however it has made the FK relationship non-nullable therefore producing bad migrations. Which I see in the comment in `AbstractContext` is desired but isn't possible with sqlite migrations, so for now I have changed it to `public int? PackageKey { get; set; }` which then does not produce any changes in the migration.

This is only for the PackageDependency entity right? Package type and the target framework entities aren’t affected by this issue and we can add the PackageKey property easily? Maybe we should fix just those two for now and fix package dependency later?

> Based on this [discussion in the ef repo](https://github.com/dotnet/efcore/issues/329#issuecomment-276138743) I am trying to produce a migration to copy the PackageDependencies data to a new table with the new non-nullable FK, but that might be better as a separate issue?

That’s a really good find, I didn’t know about this work around. I agree that’d be best in a separate issue. 

By the way, I’m out of town until Tuesday. I may be slow to respond, sorry about that!

**Comment by MarkZither** [comment]
> This is only for the PackageDependency entity right? Package type and the target framework entities aren’t affected by this issue and we can add the PackageKey property easily? Maybe we should fix just those two for now and fix package dependency later?

Thats right, Package Type and Target Framework have the PackageKey added as `int` as you proposed.

**Comment by loic-sharma** [comment]
> Based on this [discussion in the ef repo](https://github.com/dotnet/efcore/issues/329#issuecomment-276138743) I am trying to produce a migration to copy the PackageDependencies data to a new table with the new non-nullable FK, but that might be better as a separate issue?

I ran into a similar problem with the case sensitivity on the `Version` column on SQLite (see https://github.com/loic-sharma/BaGet/pull/466). SQLite cannot alter existing columns, so fixing this in a migration would require a similar trick (create new table -> copy data -> delete original table -> rename new table). I ended up doing a cheaper hack instead: I fixed the column by modifying the initial migration instead of adding a new migration. This fixes new databases going forward; old databases don't get the fix but can be recreated. I think we should take a similar approach for the `PackageDependencies` table.
